### PR TITLE
Specify the block that PtGradientCollector belongs to

### PR DESCRIPTION
**Pull request by KexinFeng**
The zeroGradients() call (added in https://github.com/deepjavalibrary/djl/pull/2101 to solve https://github.com/deepjavalibrary/djl/issues/2024) when constructing PtGradientCollector() compromises the efficiency and memory usage, as shown in https://github.com/deepjavalibrary/djl/issues/2210. Thus the block has to be specified so that the zeroGradients is called only on its parameters, and the search scope is narrowed down. Another relevant PR is https://github.com/deepjavalibrary/djl/pull/2111. 

The original PtGradientCollector() is left unchanged in order to be backward compatible. A new PtGradientCollector(Block) is added so that the gradient collector is restricted to operate on a certain block. Then multiple such collectors can coexist. @zachgk I added this point to the doc, please take a look. The relevant PR https://github.com/deepjavalibrary/djl/pull/2111.

**In essence, this PR is to disable `zeroGradients()` in the default case, unless the block is specified when creating a GradientCollector.** This is because of the issues inside global search in zeroGradients.

---
To solve this issue the issue https://github.com/deepjavalibrary/djl/issues/2024 in the Dive into Deep Learning book, how the gradient is accumulated should be specified in the future, either "write" or "add". Also "retain_graph" should also be fixed. Under certain condition error should be thrown, for example, if retain_graph=false and repetative backward is called. This way the ambiguity in the defination of repetative backward call, as shown in https://github.com/deepjavalibrary/djl/issues/2024, will be clarified.

On pytorch, by default gradients are added for repetative backward call and if retain_graph=true
```python
    x1 = pt.tensor([1., 2., 3., 4.], requires_grad=True)
    x2 = pt.tensor([5., 6., 7., 8.], requires_grad=True)
    u = x1 * x2
    z = u * x1
    z.sum().backward(retain_graph=True)
    print("\n")
    print(x1.grad, x2.grad)

    z.sum().backward(retain_graph=True)
    print(x1.grad, x2.grad)

    z.sum().backward()
    print(x1.grad, x2.grad)
```
Output:
```
tensor([10., 24., 42., 64.]) tensor([ 1.,  4.,  9., 16.])
tensor([ 20.,  48.,  84., 128.]) tensor([ 2.,  8., 18., 32.])
tensor([ 30.,  72., 126., 192.]) tensor([ 3., 12., 27., 48.])
```
---
Note:
* If repetative backward needs to be called, `retain_graph=True` should be set. This is not an commenly seen usecase in practice. See also https://github.com/deepjavalibrary/djl/pull/2111.
* `Block` currently does not have a direct call feature like the following but requires `parameterStore`.
```python
m = nn.Linear(20, 30)
input = torch.randn(128, 20)
output = m(input)
```
---
Relavant stackoverflow issue:
https://stackoverflow.com/questions/70998652/djl-gradientcollector-try-with-resources-initialiser-error

### CODE REVIEW

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtGradientCollector.java [hunk 1]
```diff
@@ -37,8 +39,18 @@ public PtGradientCollector() {
                     "A PtGradientCollector is already collecting. Only one can be collecting at a"
                             + " time");
         }
+    }
 
-        zeroGradients();
```
Comment (zachgk) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtGradientCollector.java:41]: I think we still want the zeroGradients here. Otherwise, it won't work correctly for cases without a block

File: api/src/main/java/ai/djl/training/GradientCollector.java [hunk 2]
```diff
@@ -28,7 +28,8 @@
  * and added together.
  *
  * <p>Due to limitations in most engines, the gradient collectors are global. This means that only
- * one can be used at a time. If multiple are opened, an error will be thrown.
+ * one can be used at a time. If multiple are opened, an error will be thrown. However, if the
+ * collector is specified to operate on a certain block, then multiple collectors can coexist.
```
Comment (zachgk) [inline api/src/main/java/ai/djl/training/GradientCollector.java:32]: You can't actually have multiple (even on separate blocks) based on the engine architectures. Maybe in a future engine or if they make the training mode non-global. Or, we can just treat that as an engine limitation and then throw `UnsupportedOperationExceptions` while the test suite should expect it to work rather than expect it to fail.

File: api/src/main/java/ai/djl/engine/Engine.java [hunk 3]
```diff
@@ -306,6 +307,14 @@ public int getGpuCount() {
...
+    /**
+     * Returns a new instance of {@link GradientCollector}.
+     *
+     * @param block the block that the gradient collector attaches
+     * @return a new instance of {@link GradientCollector}
+     */
+    public abstract GradientCollector newGradientCollector(Block block);
+
     /**
      * Returns a new instance of {@link ParameterServer}.
```
Comment (zachgk) [inline api/src/main/java/ai/djl/engine/Engine.java:319]: Can you also update `trainer.newGradientCollector()` to use this? You can change it's definition to:

```java
    public GradientCollector newGradientCollector() {
        return manager.getEngine().newGradientCollector(getModel().getBlock());
    }
```

Comment (KexinFeng) [reply api/src/main/java/ai/djl/engine/Engine.java:319]: The tricky part is that inside the newGradientCollector(), the zeroGradient is not called, while inside newGradientCollector(Block block), it is. This is to avoid the efficiency and memory leak issue with global search scope as discussed  last week.

File: api/src/main/java/ai/djl/training/GradientCollector.java (hunk 2 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/training/GradientCollector.java:32]: Got it. I'll remove it then. Or maybe we can put it to TODO. Here is a relavant usecase: https://stackoverflow.com/questions/70998652/djl-gradientcollector-try-with-resources-initialiser-error

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtGradientCollector.java (hunk 1 above)
Comment (KexinFeng) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtGradientCollector.java:41]: The reason of removing it was mainly due to efficiency and memory issue https://github.com/deepjavalibrary/djl/issues/2210.
Should we add this back and then do
```
    public GradientCollector newGradientCollector() {
        return manager.getEngine().newGradientCollector(getModel().getBlock());
    }
```
Does this avoid the global search scope issue?

File: integration/src/main/java/ai/djl/integration/tests/training/GradientCollectorIntegrationTest.java [hunk 4]
```diff
@@ -107,16 +108,25 @@ public void testZeroGradients() {
...
+            block.initialize(manager, DataType.FLOAT32, inputShape);
+            ParameterList list = block.getParameters();
+            NDArray weight = list.get("weight").getArray();
+            NDArray bias = list.get("bias").getArray();
 
             for (int i = 0; i < 3; i++) {
-                try (GradientCollector gc = Engine.getInstance().newGradientCollector()) {
-                    NDArray b = a.mul(2);
-                    gc.backward(b);
+                try (GradientCollector gc = Engine.getInstance().newGradientCollector(block)) {
```
[the remark this answers is not in the document]
Comment (KexinFeng) [reply integration/src/main/java/ai/djl/integration/tests/training/GradientCollectorIntegrationTest.java:125]: The test of newGradientCollector(block) has been covered in `testClearGradients`. Do we need to add it or do we change the name of this test to be, e.g. testNewGradientCollectorWithClearGradients?

File: api/src/main/java/ai/djl/engine/Engine.java (hunk 3 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/engine/Engine.java:319]: Another difference is 
```java
        boolean wasCollecting = isCollecting.getAndSet(true);
        if (wasCollecting) {
            throw new IllegalStateException(
                    "A PtGradientCollector is already collecting. Only one can be collecting at a"
                            + " time");
        }
```
is not called inside `PtGradientCollector(Block block)`.

File: api/src/main/java/ai/djl/training/GradientCollector.java (hunk 2 above)
Comment (KexinFeng) [reply api/src/main/java/ai/djl/training/GradientCollector.java:32]: Here I was guessing: inside `PtGradientCollector(Block block)`, if the following is not called unlike that in `PtGradientCollector()`, would it allow multiple GradientCollectors to coexist?
```java
        boolean wasCollecting = isCollecting.getAndSet(true);
        if (wasCollecting) {
            throw new IllegalStateException(
                    "A PtGradientCollector is already collecting. Only one can be collecting at a"
                            + " time");
        }
```

File: api/src/main/java/ai/djl/engine/Engine.java (hunk 3 above)
Comment (enpasos) [reply api/src/main/java/ai/djl/engine/Engine.java:319]: In `zeroGradients` `array.getGradient().subi(array.getGradient());` is still used. 
Couldn't it be replaced with `JniUtils.zeroGrad((PtNDArray) array);` as used in `zeroGradients(Block block)`?

Comment (KexinFeng) [reply api/src/main/java/ai/djl/engine/Engine.java:319]: > As @zachgk mentioned ... in case you go with the approach it must also work when using a trainer.

That would be ideal. But it has the efficient and memory issue for now. So we should disable it for now.

> Are you sure that you get all relevant NDArrays with Gradient by getting the parameter NDArrays from block?

The gradientCollector is specified to a block. By defination it does search only within a block.
